### PR TITLE
Fix number of quadrature points

### LINKED ISSUE #139: [master] Overhaul of assembly methods
- Fix treatment of source term
- Fix treatment of boundary integrals
- Differentiate assembly in case of coupled or uncoupled problems

### PR DESCRIPTION
**Pull request by sebglane**

### CODE REVIEW

File: source/convection_diffusion/assemble_constant_matrices.cc [hunk 1]
```diff
@@ -19,11 +19,8 @@ void HeatEquation<dim>::assemble_constant_matrices()
   mass_matrix       = 0.;
   stiffness_matrix  = 0.;
 
-  // Compute the highest polynomial degree from all the integrands
-  const int p_degree = 2 * temperature->fe_degree;
-
   // Initiate the quadrature formula for exact numerical integration
-  const QGauss<dim>   quadrature_formula(std::ceil(0.5 * double(p_degree + 1)));
+  const QGauss<dim>   quadrature_formula(temperature->fe_degree);
```
Comment (sebglane) [inline source/convection_diffusion/assemble_constant_matrices.cc:23]: This is wrong

File: source/navier_stokes_projection/assemble_mass_and_laplace_matrices.cc [hunk 2]
```diff
@@ -17,13 +17,10 @@ void NavierStokesProjection<dim>::assemble_velocity_matrices()
   velocity_mass_matrix    = 0.;
   velocity_laplace_matrix = 0.;
 
-  // Compute the highest polynomial degree from all the integrands
-  const int p_degree = 2 * velocity->fe_degree;
-
   // Initiate the quadrature formula for exact numerical integration
-  const QGauss<dim>   quadrature_formula(std::ceil(0.5 * double(p_degree + 1)));
+  const QGauss<dim>   quadrature_formula(velocity->fe_degree);
```
Comment (sebglane) [inline source/navier_stokes_projection/assemble_mass_and_laplace_matrices.cc:21]: This is wrong

**Comment by sebglane** [comment]
Ready for review. The output of the DFG benchmark test were slightly adjusted because of differences in the numerical values.

File: source/convection_diffusion/assemble_constant_matrices.cc (hunk 1 above)
Comment (j507) [reply source/convection_diffusion/assemble_constant_matrices.cc]: How is this wrong? In this method the mass matrix and the stiffness matrix are assembled simultaneously. As such, the highest polynomial is two times the polynomial degree of the Lagrange element. The `std::ceil` call here is based in the formula to assured that the quadrature rule used in the numerical integration produces the same result as its analytical integration.

File: source/navier_stokes_projection/assemble_mass_and_laplace_matrices.cc (hunk 2 above)
Comment (j507) [reply source/navier_stokes_projection/assemble_mass_and_laplace_matrices.cc]: Same as above.

File: source/convection_diffusion/assemble_constant_matrices.cc (hunk 1 above)
Comment (sebglane) [reply source/convection_diffusion/assemble_constant_matrices.cc]: These comments were just a note for me that
```c++
   const QGauss<dim>   quadrature_formula(temperature->fe_degree); 
```
is wrong. The comment is outdated.

**Comment by j507** [comment]
What is the motivation behind deleting all the formulas to determine the exact quadrature rule for all the assembly?

**Comment by sebglane** [comment]
> What is the motivation behind deleting all the formulas to determine the exact quadrature rule for all the assembly?

I want to have a quadrature degree. The reason is that this makes sure that the same degree in every problem, regardless whether a boundary or source term is present. I hope that this makes results reproducible.

**Comment by j507** [comment]
> > What is the motivation behind deleting all the formulas to determine the exact quadrature rule for all the assembly?
> 
> I want to have a quadrature degree. The reason is that this makes sure that the same degree in every problem, regardless whether a boundary or source term is present. I hope that this makes results reproducible.

Even if I can not showcase numerical examples against the coded quadrature rules, I am somewhat adverse to this approach. While an exact integration is not exactly needed for satisfactory accurate results, I feel that this over simplification could lead to wrong results.

**Comment by sebglane** [comment]
> Even if I can not showcase numerical examples against the coded quadrature rules, I am somewhat adverse to this approach. While an exact integration is not exactly needed for satisfactory accurate results, I feel that this over simplification could lead to wrong results.

Well, for a standard problem without body forces and without boundary terms, the quadrature formulas are exact for Q2-Q1 element pair. I don't expect that the change dramatically influence the results. The wrong initialization discussed [here](https://github.com/j507/RotatingMHD/pull/136#issuecomment-901014027) was much more severe than the ones here.

**Review by j507 — APPROVED** [review]

**Pull request merged**